src/moonlink/src/error.rs

```diff
@@ -1,109 +1,280 @@
+        self
+    }
+}
+
+impl error::Error for ErrorStruct {
+    /// Returns the underlying source error for accessing structured information.
+    ///
+    /// # Example
+    /// ```ignore
+    /// if let Some(source) = error_struct.source() {
+    ///     if let Some(io_err) = source.downcast_ref::<std::io::Error>() {
+    ///         println!("IO error kind: {:?}", io_err.kind());
+    ///     }
+    /// }
+    /// ```
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.source.as_ref().map(|arc| arc.as_ref().as_ref())
+    }
+}
+
 /// Custom error type for moonlink
 #[derive(Clone, Debug, Error)]
 pub enum Error {
-    #[error("Arrow error: {source}")]
-    Arrow { source: Arc<ArrowError> },
+    #[error("{0}")]
+    Arrow(ErrorStruct),
 
-    #[error("IO error: {source}")]
-    Io { source: Arc<io::Error> },
+    #[error("{0}")]
+    Io(ErrorStruct),
 
-    #[error("Parquet error: {source}")]
-    Parquet { source: Arc<ParquetError> },
+    #[error("{0}")]
+    Parquet(ErrorStruct),
 
     #[error("Transaction {0} not found")]
     TransactionNotFound(u32),
 
-    #[error("Watch channel receiver error: {source}")]
-    WatchChannelRecvError {
-        #[source]
-        source: watch::error::RecvError,
-    },
+    #[error("{0}")]
+    WatchChannelRecvError(ErrorStruct),
 
     #[error("Tokio join error: {0}. This typically occurs when a spawned task fails to complete successfully. Check the task's execution or panic status for more details.")]
     TokioJoinError(String),
 
-    #[error("Iceberg error: {source}")]
-    IcebergError { source: Arc<IcebergError> },
+    #[error("{0}")]
+    IcebergError(ErrorStruct),
 
     #[error("Iceberg error: {0}")]
     IcebergMessage(String),
 
-    #[error("OpenDAL error: {source}")]
-    OpenDal { source: Arc<opendal::Error> },
+    #[error("{0}")]
+    OpenDal(ErrorStruct),
 
     #[error("UTF-8 conversion error: {0}")]
     Utf8(#[from] std::string::FromUtf8Error),
 
-    #[error("Join error: {source}")]
-    JoinError { source: Arc<tokio::task::JoinError> },
+    #[error("{0}")]
+    JoinError(ErrorStruct),
 
-    #[error("JSON serialization/deserialization error: {source}")]
-    Json { source: Arc<serde_json::Error> },
+    #[error("{0}")]
+    Json(ErrorStruct),
 }
 
 pub type Result<T> = result::Result<T, Error>;
 
+impl From<watch::error::RecvError> for Error {
+    fn from(source: watch::error::RecvError) -> Self {
+        Error::WatchChannelRecvError(ErrorStruct {
+            message: format!("Watch channel receiver error: {source}"),
+            status: ErrorStatus::Permanent,
+            source: Some(Arc::new(source.into())),
+        })
+    }
+}
+
 impl From<ArrowError> for Error {
     fn from(source: ArrowError) -> Self {
-        Error::Arrow {
-            source: Arc::new(source),
-        }
+        let status = match source {
+            ArrowError::MemoryError(_) | ArrowError::IoError(_, _) => ErrorStatus::Temporary,
```

Comment:
I quickly checked, memory error is used in two places:
- memory alignment error: https://github.com/apache/arrow-rs/blob/554cafa140f0e4a007aa00a411d3f2b63bc6076c/arrow-ipc/src/writer.rs#L1250, which is not retryable
- Indicates ENOMEM: https://github.com/apache/arrow-rs/blob/554cafa140f0e4a007aa00a411d3f2b63bc6076c/arrow-array/src/ffi_stream.rs#L272, which I wouldn't retry